src/charm.py

```diff
@@ -362,20 +185,68 @@ def _on_config_changed(self, event: EventBase) -> None:
                 self.on[f"{self.restart.name}"].acquire_lock.emit()
 
         # update client_properties whenever possible
-        self.kafka_config.set_client_properties()
+        self.config_manager.set_client_properties()
 
         # If Kafka is related to client charms, update their information.
         if self.model.relations.get(REL_NAME, None) and self.unit.is_leader():
             self.provider.update_connection_info()
 
+    def _on_update_status(self, event: EventBase) -> None:
+        """Handler for `update-status` events."""
+        if not self.healthy or not self.upgrade.idle:
+            return
+
+        if not self.state.zookeeper.broker_active():
+            self._set_status(Status.ZK_NOT_CONNECTED)
+            return
+
+        # NOTE for situations like IP change and late integration with rack-awareness charm.
+        # If properties have changed, the broker will restart.
+        self._on_config_changed(event)
+
+        try:
+            if not self.health.machine_configured():
+                self._set_status(Status.SYSCONF_NOT_OPTIMAL)
+                return
+        except SnapError:
+            self._set_status(Status.SNAP_NOT_RUNNING)
+            return
+
+        self._set_status(Status.ACTIVE)
+
+    def _on_remove(self, _) -> None:
+        """Handler for stop."""
+        self.sysctl_config.remove()
+
+    def _on_storage_attached(self, event: StorageAttachedEvent) -> None:
+        """Handler for `storage_attached` events."""
+        # new dirs won't be used until topic partitions are assigned to it
+        # either automatically for new topics, or manually for existing
+        # set status only for running services, not on startup
+        if self.workload.active():
+            self._set_status(Status.ADDED_STORAGE)
+            self.workload.set_snap_ownership(path=self.workload.paths.data_path)
+            self.workload.set_snap_mode_bits(path=self.workload.paths.data_path)
+            self._on_config_changed(event)
+
+    def _on_storage_detaching(self, event: StorageDetachingEvent) -> None:
+        """Handler for `storage_detaching` events."""
+        # in the case where there may be replication recovery may be possible
+        if self.state.peer_relation and len(self.state.peer_relation.units):
```

Comment:
I'm actually wondering about this `if`. Since it calls `on_config_changed` right after, it will replace instantly the unit status with either active or other of the blocked/waiting ones.